sstables: prepare for uuid-based generation_type

bhalevy opened the pull request:
Preparing for #10459, this series defines sstables::generation_type::int_t
as `int64_t` at the moment and use that instead of naked `int64_t` variables
so it can be changed in the future to hold e.g. a `std::variant<int64_t, sstables::generation_id>`.

sstables::new_generation was defined to generation new, unique generations.
Currently it is based on incrementing a counter, but it can be extended in the future
to manufacture UUIDs.

The unit tests are cleaned up in this series to minimize their dependency on numeric generations.
Basically, they should be used for loading sstables with hard coded generation numbers stored under `test/resource/sstables`.

For all the rest, the tests should use existing and mechanisms introduced in this series such as generation_factory, sst_factory and smart make_sstable methods in sstable_test_env and table_for_tests to generate new sstables with a unique generation, and use the abstract sst->generation() method to get their generation if needed, without resorting the the actual value it may hold.

tchaikov commented on sstables/generation_type.hh:
```diff
@@ -31,6 +31,17 @@ public:
+        return *this;
+    }
+
+    generation_type operator++(int) noexcept {
```
makes sense to be. as the random uuid might not be completely random. in the case of uuid v1, we might need to the increment the "clock_seq" bits to differentiate the new one from the latest one. so we will have to track the latest generation somewhere. put in other way, the generator of generation could be stateful.

bhalevy replied:
What do you mean by not be completely random?
`clock_seq_and_node` is based on the hw rng and unique shard id so it should be globally unique,
then get_time_UUID -> create_time_safe ensures uniqueness of the timestamp part.

tchaikov replied:
i see. if we use `UUID_gen` for generating the time uuid, the uniqueness is guaranteed. as it tracks the `_last_used_time`. so the only reason to have `operator++` is just to encapsulate the difference of these two implementations, even the latter does not really depends on the state tracked by the "last generation" -- it has a thread local `UUID_gen::_instance` for this purpose.

raphaelsc replied:
> What do you mean by not be completely random? `clock_seq_and_node` is based on the hw rng and unique shard id so it should be globally unique, then get_time_UUID -> create_time_safe ensures uniqueness of the timestamp part.

just a small correction: in clock_seq_and_node, the clock seq is a result of pnrng (with rand_r, therefore not hw rng) using time since epoch as its seed. the node is the MAC address, tampered with shard id.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4609/

avikivity commented on sstables/generation_type.hh:
```diff
@@ -22,7 +22,9 @@ public:
-    generation_type() = delete;
+    generation_type() noexcept
+        : _value(0)
+    {}
```
Seems wrong. Why would we want the default constructor to generate a random ID? Just like the default constructor for int64_t doesn't generate a random int.

Creating a new generation should be explicit.

denesb replied:
I agree, this is too subtle. Our UUID:s also work this way: default ctor creates null UUID, and we have a factory method to generate a random one.

tchaikov replied:
we need a default ctor so we are able to accept `generation_type` as an boost::program::option, which uses `boost::lexical_cast<T>()` to cast a `std::string` to the given type. see [`tools/scylla-sstable.cc`](https://github.com/scylladb/scylladb/blob/fb898d214c5aa36689c941b015466a662471f52e/tools/scylla-sstable.cc#L2468) after Benny's change of https://github.com/scylladb/scylladb/pull/12994/commits/5764aed61ad9438201ac4cfd0fc73c9a17ac50a6#diff-4cd1a5351a4e8c9836e4cdf6eaa67498e53d365fe9ab76dcae4baa9e6ac4cdc0R2469. despite that this change has not been landed. i think the change makes sense in general. 

probably once we have UUID based generation id, we will need to use `std::variant<std::monostate, int64_t, UUID>`. otherwise we should have `std::variant<UUID, int64_t>`. as, at least, we can tell if the uuid is generated with the value-initialized ctor just like `UUID{}`. 

or we will need to provide a customized validator for this option.

denesb replied:
I'm not saying we shouldn't have a default constructor, just that it should create a "null" UUID, not a random one.

tchaikov replied:
ahh, i see. it makes more sense.

avikivity commented on sstables/generation_type.hh:
```diff
@@ -31,6 +31,17 @@ public:
+    generation_type& operator++() noexcept {
+        ++_value;
+        return *this;
+    }
```
I'd be happier with a method new_generation somewhere (not in generation_type). Note it depends on the table's settings, not just on what the latest generation looks like.

tchaikov replied:
yeah, i thought about this also. it's like a move-only input iterator -- we can only read from it for a new value. the `value_type` of this iterator is `generation_type`. but they are not of the same type.

raphaelsc replied:
Yes, generation_type should be restricted to storing info about a single generation. All the coordination should be performed in replica::table, possibly with some guidance from sstable_manager (as to what generation type should be used). sstable_manager also guides replica::table as to what is the highest sstable version supported.

avikivity commented on replica/database.hh:
```diff
@@ -424,6 +424,11 @@ private:
+    // generation number in the table so that newly created sstables
+    // will use numeric generation numbers larger than this one.
+    // Initialized when the table is populated via update_sstables_known_generation.
     std::optional<int64_t> _sstable_generation = {};
```
This should be replaced with a new_generation() method, that either creates a new random ID or uses _sstable_generation to pick up a new one. We shouldn't call operator++() on something that isn't an integer to do an operation that isn't incrementing something.

tchaikov replied:
is it something like #13073 ?

avikivity replied:
Yes, but that looks like overkill to me.

denesb commented on test/boost/reader_concurrency_semaphore_test.cc:
```diff
@@ -1356,11 +1356,13 @@ memory_limit_table create_memory_limit_table(cql_test_env& env, uint64_t target_
             while (num_sstables != target_num_sstables) {
                 ++num_sstables;
-                auto sst = sst_man.make_sstable(s, sstables_dir.path().string(), sstables::generation_type{num_sstables}, sst_man.get_highest_supported_format(), sstables::sstable_format_types::big);
+                ++gen;
```
I think this increment syntax works well only with the current integer based generation, which we want to move away from. So let's make this an explicit method instead: `auto gen = generation_type::generate_random_gen();`
 (name can be something else).

denesb replied:
We might even rename `generation -> id` to emphasize that this is not a simple incremented integer anymore: it is a unique id, whose actual value has no meaning.
Might not worth the churn though.

tchaikov replied:
generation is encoded using an uuid, maybe we can keep this naming legacy? i admit this is not ideal though. but id is way too general. but generation is what we've been using. just like we are still using "disk" even if we are actually talking about is NAND flash memory. strictly speaking, it's wrong. but it's more like a metaphor.

raphaelsc replied:
Botond is right. Increment syntax would only make sense if we'd enforce a monotonically increasing requirement in the interface. Like it was discussed, sstables::generation_type shouldn't be given the responsibility of generating a new value from the current one.

tchaikov commented:
@bhalevy hi Benny, are you still working on revising this changeset? probably i can help a little bit on this by sending a PR to your branch? two changes in my mind are

- use `generation_type` when appropriate, for instance, https://github.com/scylladb/scylladb/pull/13074/commits/077286645736ca78777cf8f797addd70a01955c6#diff-e77ac6a4425e9a7eec3d461a5896cc62a6fcca38b02ad855e047f9baace8b5b3R963
- drop the incremental operator. i will try to come up a simpler version later.
- drop the default constructor. once we use `std::variant<>` for the value of the generation , we should be able to have a better default value then.

what do you think?

bhalevy commented:
> @bhalevy hi Benny, are you still working on revising this changeset? probably i can help a little bit on this by sending a PR to your branch? three changes in my mind are

I'm close to sending a new revision

> 
> * use `generation_type` when appropriate, for instance, [0772866#diff-e77ac6a4425e9a7eec3d461a5896cc62a6fcca38b02ad855e047f9baace8b5b3R963](https://github.com/scylladb/scylladb/commit/077286645736ca78777cf8f797addd70a01955c6#diff-e77ac6a4425e9a7eec3d461a5896cc62a6fcca38b02ad855e047f9baace8b5b3R963)

yes, working on it.

> * drop the incremental operator. i will try to come up a simpler version later.

dropped in next version

> * drop the default constructor. once we use `std::variant<>` for the value of the generation , we should be able to have a better default value then.

dropped in next version

> 
> what do you think?

good suggestions. please hold your breath for a little longer :)

tchaikov commented:
thank you! Benny. sure =)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4890/

tchaikov commented on sstables/generation_type.hh:
```diff
@@ -49,6 +49,26 @@ inline generation_type new_generation(std::optional<generation_type> prev = std:
+        _next = new_generation(_next);
+        return *_prev;
+    }
+};
```
just a head-up, this is likely to conflict with the change in https://github.com/scylladb/scylladb/pull/13073/commits/1aa802f8c5a5e9175b5a2452691003b8644eea0f .

bhalevy replied:
I suggest this PR should be merged first as it's likely to bit-rot otherwise, and it provides some basic functionality that can be used by later changes to the product code.

tchaikov replied:
absolutely. i will take another look at this PR. my changes are based on this changeset. the early we have this PR landed the better.

avikivity replied:
Again I don't see why we need a generation_factory for something so simple. Can't it just be part of the code that gets a new generation?

bhalevy replied:
Unfortunately we have several places that generate new generations in the code, especially in tests.
I wanted whoever needs a new generation to get one in abstract way, not caring if the factory dispatches integer generations or a uuid ones.
With such a class we can also specialize it to add smp::count to the internal counter for table shards (until uuid generations)

bhalevy commented:
In v3 ([8176b84](https://github.com/scylladb/scylladb/commit/8176b84064ff08821ec7a90c37f70e7909dbd06e)):
- fixed typo in commit shortlog
- fixed use of sst_factory in mutation_reader_test.do_test_clustering_order_merger_sstable_set
- sstable_directory_test_generation_sanity:
  - run per-shard sstdir.do_for_each_sstable n a seastar::async thread.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4893/

bhalevy commented:
Conflicts now with https://github.com/scylladb/scylladb/pull/13116

bhalevy commented:
In v4 ([04ff168](https://github.com/scylladb/scylladb/commit/04ff1687072f348f19f0ef292f0a36e43b35b480)):
- rebased on top 1379d8330f1966fec818cfe5567544091bbfdfd2
- Added 'test: sstable_compaction_test: use cf.make_sstable to abstract sstable generations'

bhalevy commented:
In v5 ([fb151ba](https://github.com/scylladb/scylladb/commit/fb151ba9868a1ecdb0a72d2e06bc6b316883ad67)):
- restored `sstable_version_types::me` to sstable_3_x_test
  - I initially thought it was required to be removed, but apparently I had a bug and it is not required.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4948/

denesb commented:
What is this PR's relation with https://github.com/scylladb/scylladb/pull/13073? They conflict with each other. Should the former go in first?

bhalevy commented:
https://github.com/scylladb/scylladb/pull/13073 should be based after this one
as this PR prepares the testing infrastructure for later changes in the "production" stack

tchaikov commented on test/boost/sstable_resharding_test.cc:
```diff
@@ -43,13 +43,13 @@ void run_sstable_resharding_test() {
 
     // create sst shared by all shards
-    {
+    sstables::shared_sstable sst = [&] {
```
nit, use `std::invoke()` for better readability, so the human reader can understand this is a function call not a definition of a lambda. i learned this from a comment from Avi.

bhalevy replied:
will fix in next version

bhalevy commented on replica/distributed_loader.cc:
```diff
@@ -556,7 +549,7 @@ class table_populator {
     std::unordered_map<sstring, lw_shared_ptr<sharded<sstables::sstable_directory>>> _sstable_directories;
     sstables::sstable_version_types _highest_version = sstables::oldest_writable_sstable_format;
-    sstables::generation_type _highest_generation = sstables::generation_from_value(0);
+    std::optional<sstables::generation_type> _highest_generation;
```
the default ctor for generation_type is deleted in this branch, but it makes sense to define it as you describe above, like a disengaged optional.

tchaikov replied:
yeah, will add in in my change based on yours.

bhalevy commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -211,25 +211,30 @@ SEASTAR_TEST_CASE(compact) {
                         BOOST_REQUIRE(m->key().equal(*s, partition_key::from_singular(*s, data_value(sstring("jerry")))));
                         BOOST_REQUIRE(!m->partition().partition_tombstone());
-                        auto rows = m->partition().clustered_rows();
+                        const auto& rows = m->partition().clustered_rows();
```
Previously, the code used a mutation mutation, so `clustered_rows()` resolved to
https://github.com/scylladb/scylladb/blob/1d9b7f3a929bb29454f3f5152785f920f3119fe1/mutation/mutation_partition.hh#L1426
while with `verify_mutation = [&] (std::function<void(const mutation_opt&)>` it resolves to
https://github.com/scylladb/scylladb/blob/1d9b7f3a929bb29454f3f5152785f920f3119fe1/mutation/mutation_partition.hh#L1425

bhalevy replied:
I should have broken the patch respectively.
It's crammed with the async thread change, but it is orthogonal indeed.

tchaikov replied:
ahh, i see now. thanks!

bhalevy replied:
That said, I'm preparing a new version that passes the mutation_opt by value to the verifier function to reduce the churn (and the `verify_mutation` wrapper does nothing further with the mutation_opt so it can just as well pass it on to the verify function)

denesb commented on replica/database.hh:
```diff
@@ -1092,6 +1093,7 @@ public:
     friend std::ostream& operator<<(std::ostream& out, const column_family& cf);
     // Testing purposes.
     friend class ::column_family_test;
+    friend class ::table_for_tests;
```
Why this friendship? I only see public methods used below.

denesb replied:
AFAIK https://github.com/scylladb/scylladb/pull/13073 removes this so, when we merge that we should remember to remove this friendship.

bhalevy replied:
adding comment in next version

denesb commented on test/lib/sstable_test_env.hh:
```diff
@@ -85,12 +85,19 @@ public:
         return make_sstable(std::move(schema), _impl->dir.path().native(), generation, std::move(v), std::move(f), buffer_size, now);
     }
 
+    [[deprecated("use sstables::generation_type param")]]
```
Why not just clean up all users? Are you planning to do that later?

denesb replied:
ok

bhalevy replied:
adding note in next version

xemul commented on sstables/generation_type.hh:
```diff
@@ -18,21 +18,26 @@
 class generation_type {
-    int64_t _value;
+public:
+    using int_t = int64_t;
```
nit: Since the intention is to 

> it can be used everywhere to prepare for uuid sstable generation support.

then maybe it's worth giving it a `value_t` name from the very beginning? Other wise uuid-switching PR would need to `s/int_t/something_else_t/`

tchaikov replied:
i think `int_t` only matters where we expect a integer-like identifier, and we want to hide the exact integer type from user. in the uuid-switching PR, i think we will have a simpler way for accessing the underlying integer identifier, either use a `explicit operator int64_t()` or `get<int64_t>()`. IMHO, `int_t` is an intermediate solution before we have uuid-based generation identifier.

bhalevy replied:
I tried to get rid of it as much as possible in this series by converting to abstract `sstables::generation_type`
those place that are left deal only with the `int_t`.
With UUID support I imagine that the `value_type` would be a `std::variant<int_t, uuid_t>`

avikivity replied:
value_type is the common name for this type of thing.

bhalevy replied:
Maybe it could be gone once when we finish the transition to UUID, but I'd still like places that use a numeric generation to use a typedef, not plain int64_t/unsigned long/long/unsigned type(s) like we use ad-hoc today all over the place.

bhalevy replied:
I'm contemplating even to use a strong type based on the tagged_integral class I proposed in https://github.com/scylladb/scylladb/pull/12966/commits/abb832238c015ab7428d95d363f058a923c7e352

xemul commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -211,21 +211,25 @@ SEASTAR_TEST_CASE(compact) {
+                    auto verify_mutation = [&] (std::function<void(mutation_opt)> verify) {
+                        std::invoke(verify, read_mutation_from_flat_mutation_reader(reader).get());
+                    };
+                    return verify_mutation([&] (mutation_opt m) {
```
Yikes, doesn't this `return` bails out and thus skipt all the subsequent `verify_mutation` calls?

bhalevy replied:
thanks eagleeye!

xemul replied:
:eagle:

bhalevy replied:
fixed in next version

xemul commented on test/boost/mutation_reader_test.cc:
```diff
@@ -3883,8 +3883,7 @@ static future<> do_test_clustering_order_merger_sstable_set(bool reversed) {
             auto sst_factory = [table_schema, &env, &tmp, gen = ++gen] () {
-                return env.make_sstable(std::move(table_schema), tmp.path().string(), gen,
-                    sstables::sstable::version_types::md, sstables::sstable::format_types::big);
+                return env.make_sstable(std::move(table_schema), tmp.path().string(), gen);
```
was: sstables::sstable::version_types::md
now: (defaulted to) sstables::get_highest_sstable_version() -> me

bhalevy replied:
yes, probably should have separated that.
the was no need to use `md` sepcifically.

bhalevy replied:
patch extracted in next version

xemul commented on test/boost/sstable_3_x_test.cc:
```diff
@@ -3023,20 +3026,18 @@ static flat_mutation_reader_v2 compacted_sstable_reader(test_env& env, schema_pt
-        return env.make_sstable(s, tmp.path().string(), new_generation,
-                         sstables::sstable_version_types::mc, sstable::format_types::big, 4096);
+    desc.creator = [s, &env, new_generation] (shard_id dummy) {
+        return env.make_sstable(s, new_generation);
```
was: sstable_version_types::mc
now: (defaulted) get_highest_sstable_version() -> me

bhalevy replied:
ditto as for use of `md` above

bhalevy replied:
patch extracted in next version

xemul commented on test/boost/sstable_3_x_test.cc:
```diff
@@ -3023,20 +3026,18 @@ static flat_mutation_reader_v2 compacted_sstable_reader(test_env& env, schema_pt
     sstables::compact_sstables(std::move(desc), cdata, cf->as_table_state()).get();
 
-    auto compacted_sst = open_sstable(env, s, tmp.path().string(), new_generation);
+    auto compacted_sst = env.reusable_sst(s, new_generation).get();
```
open_sstable() opened the `mc` version, this defaults to highest-version which is `me`

Presumably there's more of that kind below, and I admit this switch was deliberate to let test-cases test newer versions of sstables, but in that case it should be at least reflected in the commit message

xemul commented on test/boost/sstable_3_x_test.cc:
```diff
@@ -3200,55 +3203,59 @@ static sstring get_write_test_path(sstring table_name) {
+                                  "ks", table_name, sstable_version_types::mc, generation_from_value(1), big, file_type);
         auto result_filename =
-                sstable::filename(result_path.string(), "ks", table_name, version, generation_from_value(1), big, file_type);
+                sstable::filename(result_path.string(), "ks", table_name, sst->get_version(), sst->generation(), big, file_type);
```
This change doesn't fit the "use make_sstable defaults as much as possible" description

bhalevy replied:
extracted out in next version (need actually a few patch to make incremental, standalone changes)

xemul commented on replica/distributed_loader.cc:
```diff
@@ -580,7 +573,7 @@ class table_populator {
 
         co_await smp::invoke_on_all([this] {
-            _global_table->update_sstables_known_generation(_highest_generation);
+            _global_table->update_sstables_known_generation(_highest_generation.value_or(sstables::generation_type(0)));
```
Why 0? If the generation is to be set on table's sstables, shouldn't this be a "valid" generation, say (according to previous patch) the `new_generation(std::nullopt)`?

bhalevy replied:
hmm, or just pass `_highest_generation` itself (which contains an optional)
and let `update_sstables_known_generation` decide what to do about it.

bhalevy replied:
passing optional generation in next version

xemul commented on test/lib/sstable_utils.hh:
```diff
@@ -245,6 +247,65 @@ future<> for_each_sstable_version(AsyncAction action) {
+    size_t buffer_size = sstables::default_sstable_buffer_size;
+private:
+    generation_factory local_gen_factory;
+    generation_factory& gen_factory = local_gen_factory;
```
Note, that test_env::make_sstable() can already produce sstables with "monotonic" generations itself. It has the generation counter on board since be7f4ff53a73a7a0817629160a5ac7e306a1b616 . So this factory is very likely duplicating this functionality.

I'm not proposing to refactor it here though

bhalevy replied:
yeah, when I originally wrote ths code code be7f4ff53a73a7a0817629160a5ac7e306a1b616 was either in your branch or not born yet incepted

xemul commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -3866,17 +3866,14 @@ SEASTAR_TEST_CASE(test_offstrategy_sstable_compaction) {
             ss.add_row(mut, ss.make_ckey(0), "val");
 
-            auto cf = env.make_table_for_tests(s, tmp.path().string());
+            auto cf = env.make_table_for_tests(s, tmp.path().string(), version);
```
This caught my ~~eagle~~eye . Why is version argument added here? Is it runaway from one of previous patches?

bhalevy replied:
It's needed for cf.make_sstable to use that version.
I'm refactoring the patches in finer resolution in the next version so it will be more visible and better explained

xemul commented on test/lib/test_services.hh:
```diff
@@ -76,4 +79,18 @@ struct table_for_tests {
+        return [this] { return make_sstable(); };
+    }
+
+    operator std::function<sstables::shared_sstable()>() {
```
Please, add a comment that this thing is to make table_for_tests be useable as the factory argument to make_sstable_containing

bhalevy replied:
I'm actually getting rid of the cast to function in the next version, since I don't like it myself, in hindsight.
Instead, I'm introducing a `make_sst_factory` method to sstable_test_env and to table_for_tests which will be clearer as to what's going on.

xemul commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -299,30 +299,29 @@ static std::vector<sstables::shared_sstable> get_candidates_for_leveled_strategy
 static future<std::vector<sstables::generation_type::int_t>> compact_sstables(test_env& env, std::vector<sstables::generation_type::int_t> generations_to_compact,
         sstables::generation_type::int_t new_generation, bool create_sstables, uint64_t min_sstable_size, compaction_strategy_type strategy) {
     BOOST_REQUIRE(smp::count == 1);
+    return seastar::async(
```
A suggestion. There are several patches that toss seastar::async() stuff around sstable tests. I might make sense to submit them altogether as a separate PR

avikivity replied:
All the patches that clean up / refactor tests can be upstreamed quickly.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14/

avikivity commented on test/boost/mutation_reader_test.cc:
```diff
@@ -3848,7 +3848,7 @@ static future<> do_test_clustering_order_merger_sstable_set(bool reversed) {
     auto make_tested = [&env, query_schema, pk = g._pk, &pr, &query_slice, reversed]
             (const time_series_sstable_set& sst_set,
-                const std::unordered_set<int64_t>& included_gens, streamed_mutation::forwarding fwd) {
+                const std::unordered_set<sstables::generation_type::int_t>& included_gens, streamed_mutation::forwarding fwd) {
```
Why not use generation_type, here and elsewhere?

bhalevy replied:
it is converted to that later on in the series, once possible

bhalevy replied:
Also generation_type(int_t) ctor is defined as explicit, on purpose to catch inadvertent type conversion.
If we make this change early, then we need to allow implicit conversion and maybe by the end of the series make it explicit again after making all intended conversions explicit.

avikivity commented on sstables/generation_type.hh:
```diff
@@ -41,6 +41,14 @@ constexpr generation_type::int_t generation_value(generation_type generation) {
+    if (!prev) {
+        return generation_type(1);
+    } else {
+        return generation_type(prev->value() + 1);
```
Don't we usually increment by smp::count?

bhalevy replied:
We can but today that's done in replica::table.
The tests increment by one (generally speaking).

avikivity replied:
ok. But what happens to the code in replica::table? Is it still using ints?

Perhaps we should increment by smp::count here too, if the code is intended to switch over.

bhalevy commented:
Generic cleanups pushed to #13242.
Turning to draft until the it is merged.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/142/

bhalevy commented:
rebased

raphaelsc commented on replica/table.cc:
```diff
@@ -66,6 +66,33 @@ static seastar::metrics::label keyspace_label("ks");
 using namespace std::chrono_literals;
 
+sstables::generation_type table::make_new_generation(std::optional<sstables::generation_type> prev) {
+    auto prev_value = prev.value_or(sstables::generation_type(0)).value();
```
what about we merge calculate_generation_for_new_table() into this function (and then get rid of the former)? 

so we do:
auto prev_value = prev.value_or(_sstable_generation).value();

as we don't want to ignore the known generation

Also please rename make_new_generation to make_new_sstable_generation.

raphaelsc commented on sstables/sstables.cc:
```diff
@@ -2414,7 +2414,7 @@ static entry_descriptor make_entry_descriptor(sstring sstdir, sstring fname, sst
     }
-    return entry_descriptor(sstdir, ks, cf, generation_from_value(boost::lexical_cast<unsigned long>(generation)), version, format_from_string(format), sstable::component_from_sstring(version, component));
+    return entry_descriptor(sstdir, ks, cf, generation_from_value(boost::lexical_cast<sstables::generation_type::int_t>(generation)), version, format_from_string(format), sstable::component_from_sstring(version, component));
 }
```
s -> int_t -> value_t ?

bhalevy replied:
currently it's still only int.
value_type will be a std::variant of int_t and sstable_id ( a tagged uuid).
let's change it then

tchaikov replied:
yeah, we should revisit all the `int_t` then, and 

- either generalize it to `value_t`, or
- keep on using `int_t` if the use case only applies to `int_t`, or add 
- yet another case to handle the uuid case.

tchaikov replied:
ahh, quite a few of `int_t` have been replaced with `generation_type`, so 2 out of 3 action items above have been completed!

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/227/

bhalevy commented on sstables/generation_type.hh:
```diff
@@ -13,47 +13,67 @@
+template <std::ranges::range Range, typename Target = std::vector<sstables::generation_type>>
+Target generations_from_values(const Range& values) {
+    return boost::copy_range<Target>(values | boost::adaptors::transformed([] (auto value) {
+        return generation_type(value);
```
yeah, that'd work

tchaikov commented on test/boost/database_test.cc:
```diff
@@ -409,33 +409,42 @@ SEASTAR_THREAD_TEST_CASE(test_distributed_loader_with_pending_delete) {
+    std::optional<sstables::generation_type> prev_gen;
+    std::vector<sstables::generation_type> gen;
+    size_t num_gens = 9;
+    gen.reserve(num_gens);
```
nit, could use `std::generate_n()` so we don't need to leave `prev_gen` in this scope. will create a follow-up change to address this once this change is landed.

tchaikov commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/227/

i am not able to reproduce at least some of the failing test here. presumably, it was caused by a regression in dtest. since the offending commit was removed, let's retrigger the test.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/296/

tchaikov commented:
@denesb hi Botond, are we expecting a new revision from Benny ? i think this changeset has been approved. there are some nits from Raph and me though, but we can address them in a follow-up PR, right?

denesb commented:
Queued.
